machineconfiguration/v1alpha1/types_internalreleaseimage.go

```diff
@@ -0,0 +1,145 @@
+// +openshift:enable:FeatureGate=NoRegistryClusterOperations
+// +kubebuilder:metadata:labels=openshift.io/operator-managed=
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'cluster'",message="internalreleaseimage is a singleton, .metadata.name must be 'cluster'"
+
+// InternalReleaseImage is used to keep track and manage a set
+// of release bundles (OCP and OLM operators images) that are stored
+// into the control planes nodes.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type InternalReleaseImage struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of this internal release image.
+	// +required
+	Spec InternalReleaseImageSpec `json:"spec,omitzero"`
+
+	// status describes the last observed state of this internal release image.
+	// +optional
+	Status *InternalReleaseImageStatus `json:"status,omitempty,omitzero"`
+}
+
+// InternalReleaseImageSpec defines the desired state of a InternalReleaseImage.
+type InternalReleaseImageSpec struct {
+	// releases is a list of release bundle identifiers that the user wants to
+	// add/remove to/from the control plane nodes.
+	// Entries must be unique, keyed on the name field.
+	// This field can contain between 1 and 5 entries.
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=5
+	// +listType=map
+	// +listMapKey=name
+	// +required
+	Releases []InternalReleaseImageRef `json:"releases,omitempty"`
+}
+
+// InternalReleaseImageRef is used to provide a simple reference for a release
+// bundle. Currently it contains only the name field.
+type InternalReleaseImageRef struct {
+	// name indicates the desired release bundle identifier. This field is required and must be between 1 and 64 characters long.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name,omitempty"`
+}
+
+// InternalReleaseImageStatus describes the current state of a InternalReleaseImage.
+type InternalReleaseImageStatus struct {
+	// releases is a list of the release bundles currently owned and managed by the
+	// cluster, indicating that their images can be safely pulled by any cluster entity
+	// requiring them.
+	// Entries must be unique, keyed on the name field.
+	// This field can contain between 1 and 5 entries.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=5
+	// +optional
+	Releases []InternalReleaseImageStatusRef `json:"releases,omitempty"`
+}
+
+type InternalReleaseImageStatusRef struct {
```

Comment:
Is the name of this struct correct? Does it describe what it's for? Trying to work out how we got to this name

Reply:
Meant to be a reference to the status of a given release. Honestly, didn't have a better idea to avoid repeating multiple times the term `release` (maybe `InternalReleaseImageBundleStatus`). Any suggestion?